packages/next/src/server/web-server.ts

```diff
@@ -170,187 +169,144 @@ export default class NextWebServer extends BaseServer<WebServerOptions> {
-        name: '_next catchall',
-        // This path is needed because `render()` does a check for `/_next` and the calls the routing again
-        fn: async (req, res, _params, parsedUrl) => {
-          await this.render404(req, res, parsedUrl)
-          return {
-            finished: true,
-          }
-        },
-      },
-    ]
-
-    const catchAllRoute: Route = {
-      match: getPathMatch('/:path*'),
-      type: 'route',
-      matchesLocale: true,
-      name: 'Catchall render',
-      fn: async (req, res, _params, parsedUrl) => {
-        let { pathname, query } = parsedUrl
-        if (!pathname) {
-          throw new Error('pathname is undefined')
-        }
+  protected async normalizeNextData(
+    req: BaseNextRequest,
+    res: BaseNextResponse,
+    parsedUrl: NextUrlWithParsedQuery
+  ): Promise<{ finished: boolean }> {
+    const params = getPathMatch('/_next/data/:path*')(parsedUrl.pathname)
 
-        // interpolate query information into page for dynamic route
-        // so that rewritten paths are handled properly
-        const normalizedPage = this.serverOptions.webServerConfig.normalizedPage
-
-        if (pathname !== normalizedPage) {
-          pathname = normalizedPage
-
-          if (isDynamicRoute(pathname)) {
-            const routeRegex = getNamedRouteRegex(pathname, false)
-            pathname = interpolateDynamicPath(pathname, query, routeRegex)
-            normalizeVercelUrl(
-              req,
-              true,
-              Object.keys(routeRegex.routeKeys),
-              true,
-              routeRegex
-            )
-          }
-        }
+    // Make sure to 404 for /_next/data/ itself and
+    // we also want to 404 if the buildId isn't correct
+    if (!params || !params.path || params.path[0] !== this.buildId) {
+      await this.render404(req, res, parsedUrl)
+      return {
+        finished: true,
+      }
+    }
+    // remove buildId from URL
+    params.path.shift()
 
-        // next.js core assumes page path without trailing slash
-        pathname = removeTrailingSlash(pathname)
+    const lastParam = params.path[params.path.length - 1]
 
-        if (this.i18nProvider) {
-          const { detectedLocale } = await this.i18nProvider.analyze(pathname)
-          if (detectedLocale) {
-            parsedUrl.query.__nextLocale = detectedLocale
-          }
-        }
+    // show 404 if it doesn't end with .json
+    if (typeof lastParam !== 'string' || !lastParam.endsWith('.json')) {
+      await this.render404(req, res, parsedUrl)
+      return {
+        finished: true,
+      }
+    }
 
-        const bubbleNoFallback = !!query._nextBubbleNoFallback
+    // re-create page's pathname
+    let pathname = `/${params.path.join('/')}`
+    pathname = getRouteFromAssetPath(pathname, '.json')
 
-        if (isAPIRoute(pathname)) {
-          delete query._nextBubbleNoFallback
-        }
+    // ensure trailing slash is normalized per config
+    if (this.getMiddleware()) {
```

Comment:
getMiddleware() is called here and then in `this.nextConfig.i18n` again, would be worth assigning this call to a variable and reusing it below.